Important library that we use in our convolution neural network

In [0]:
import keras
from keras.models import Sequential

from keras import regularizers, optimizers
from keras.layers import Convolution2D

from keras_preprocessing.image import ImageDataGenerator
from keras.layers import Dense, Activation, Flatten, Dropout, BatchNormalization
from keras.layers import Conv2D, MaxPooling2D
from keras import regularizers, optimizers
from keras.optimizers import adam

In [0]:
import pandas as pd
from google.colab import files


library to import pandas and upload files on collab

In [0]:
upload=files.upload()

Saving dress_patterns.csv to dress_patterns.csv


In [0]:
dataset=pd.read_csv('dress_patterns.csv')

In [44]:
dataset.isnull().sum()

_unit_id               0
category               0
category:confidence    0
image_url              0
dtype: int64

here we analyse that there are no null data in our dataset

In [0]:
dataset.tail()

_unit_id   category  category:confidence  \
15697  854497210      plain               1.0000   
15698  854497281     floral               1.0000   
15699  854497512    squares               0.9230   
15700  854497595     tribal               0.5199   
15701  856388654  polka dot               0.9618   

                                               image_url  
15697  http://s3-eu-west-1.amazonaws.com/we-attribute...  
15698  http://s3-eu-west-1.amazonaws.com/we-attribute...  
15699  http://s3-eu-west-1.amazonaws.com/we-attribute...  
15700  http://s3-eu-west-1.amazonaws.com/we-attribute...  
15701  http://s3-eu-west-1.amazonaws.com/we-attribute...

In [0]:

dataset.shape

(15702, 4)

total number of rows in our datset 15702  and coloumn is 4

In [0]:
duplicateRowsDF = dataset[dataset.duplicated(['category:confidence', 'image_url'])]
len(duplicateRowsDF)

68

Here we analyse the duplicate link and their category:confidence score and we see that strength of duplicate link and category:confidence is 68. so we have to remove these rows from our dataset.

In [0]:
dataset2=dataset.drop_duplicates(subset=['category:confidence','image_url'], keep='first')

In [0]:
dataset2.shape

(15634, 4)

In [0]:
dataset2=dataset2.reset_index()

used to reset index


In [0]:
dataset2.tail()

index   _unit_id   category  category:confidence  \
15629  15695  851567928      plain               0.9232   
15630  15696  851568211     animal               0.7667   
15631  15699  854497512    squares               0.9230   
15632  15700  854497595     tribal               0.5199   
15633  15701  856388654  polka dot               0.9618   

                                               image_url  
15629  http://s3-eu-west-1.amazonaws.com/we-attribute...  
15630  http://s3-eu-west-1.amazonaws.com/we-attribute...  
15631  http://s3-eu-west-1.amazonaws.com/we-attribute...  
15632  http://s3-eu-west-1.amazonaws.com/we-attribute...  
15633  http://s3-eu-west-1.amazonaws.com/we-attribute...

After remove duplicacy in our dataset now length of our dataset is (rows=15634, coloumn=4)

In [0]:
dataset2=dataset2.drop('index', axis=1)

In [0]:
dataset2.tail()

_unit_id   category  category:confidence  \
15629  851567928      plain               0.9232   
15630  851568211     animal               0.7667   
15631  854497512    squares               0.9230   
15632  854497595     tribal               0.5199   
15633  856388654  polka dot               0.9618   

                                               image_url  
15629  http://s3-eu-west-1.amazonaws.com/we-attribute...  
15630  http://s3-eu-west-1.amazonaws.com/we-attribute...  
15631  http://s3-eu-west-1.amazonaws.com/we-attribute...  
15632  http://s3-eu-west-1.amazonaws.com/we-attribute...  
15633  http://s3-eu-west-1.amazonaws.com/we-attribute...

In [0]:
y=dataset2.iloc[:,3].values

In [0]:
print(len(y))

15634


length of our total link from which we scrape images

In [0]:
!apt-get install -y -qq software-properties-common python-software-properties module-init-tools
!add-apt-repository -y ppa:alessandro-strada/ppa 2>&1 > /dev/null
!apt-get update -qq 2>&1 > /dev/null
!apt-get -y install -qq google-drive-ocamlfuse fuse

from google.colab import auth
auth.authenticate_user()
from oauth2client.client import GoogleCredentials
creds = GoogleCredentials.get_application_default()
import getpass
!google-drive-ocamlfuse -headless -id={creds.client_id} -secret={creds.client_secret} < /dev/null 2>&1 | grep URL
vcode = getpass.getpass()
!echo {vcode} | google-drive-ocamlfuse -headless -id={creds.client_id} -secret={creds.client_secret}

E: Package 'python-software-properties' has no installation candidate
Selecting previously unselected package google-drive-ocamlfuse.
(Reading database ... 131304 files and directories currently installed.)
Preparing to unpack .../google-drive-ocamlfuse_0.7.3-0ubuntu1~ubuntu18.04.1_amd64.deb ...
Unpacking google-drive-ocamlfuse (0.7.3-0ubuntu1~ubuntu18.04.1) ...
Setting up google-drive-ocamlfuse (0.7.3-0ubuntu1~ubuntu18.04.1) ...
Processing triggers for man-db (2.8.3-2ubuntu0.1) ...
Please, open the following URL in a web browser: https://accounts.google.com/o/oauth2/auth?client_id=32555940559.apps.googleusercontent.com&redirect_uri=urn%3Aietf%3Awg%3Aoauth%3A2.0%3Aoob&scope=https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive&response_type=code&access_type=offline&approval_prompt=force
··········
Please, open the following URL in a web browser: https://accounts.google.com/o/oauth2/auth?client_id=32555940559.apps.googleusercontent.com&redirect_uri=urn%3Aietf%3Awg%3Aoauth%3A2.0%3Aoob&scope=ht

to make directory on our google drive

In [0]:
!mkdir -p colabData
!google-drive-ocamlfuse colabData

total number of files in our colabData directory in our drive

In [0]:
ls colabData

 aptrom.jpg
'AshishSaha_InternshalaResume-16 (ab5ea49d).pdf'
'AshishSaha_InternshalaResume-16 (b09298e6).pdf'
'AshishSaha_InternshalaResume-16 (b3fa884b).pdf'
 AshishSaha_InternshalaResume-16.pdf
 AshishSaha_InternshalaResume-5.pdf
 AshishSaha_InternshalaResume.pdf
 assignment.zip
 clouderizer/
'Colab Notebooks'/
'Dup(01)IMG_20141104_103348.jpg'
 EAadhaar_715553633346_16082017081428_016492.pdf
'Emailing AshishSaha_InternshalaResume-7.pdf'
'Emailing DOC-20170927-WA0003.pdf'
'Emailing INDEX mt grp1-1.pdf'
'Getting started'
 googlestockprice_regression.py
 Greendeck/
 IMG-20170109-WA0002.jpg
 IMG-20170109-WA0003.jpg
 IMG-20170109-WA0004.jpg
 IMG-20170109-WA0005.jpg
 IMG_20170110_093832.jpg
 IMG_20170110_093858.jpg
 IMG_20170110_093923.jpg
 IMG_20170110_094130.jpg
 IMG_20170110_105512.jpg
 IMG_20171031_123710.jpg
 IMG_20171031_123815.jpg
 IMG_20171031_123821.jpg
 IMG-20171219-WA0026.jpg
 IMG_20180317_123001.jpg
 IMG_20180317_123015.jpg
 IMG_20180327_172220.jpg
 IMG_20180327_172222.jpg
 IMG

In [0]:
#For fetching Dataset from link and stored in google drive
import urllib
import urllib.request
count=14986
for i in y[14986:]:
    print(count)
    resource = urllib.request.urlopen(i)
    url='colabData/Greendeck/{0}.jpg'.format(count)
    output=open(url,"wb")
    output.write(resource.read())
    output.close()
    count=count+1

14986
14987
14988
14989
14990
14991
14992
14993
14994
14995
14996
14997
14998
14999
15000
15001
15002
15003
15004
15005
15006
15007
15008
15009
15010
15011
15012
15013
15014
15015
15016
15017
15018
15019
15020
15021
15022
15023
15024
15025
15026
15027
15028
15029
15030
15031
15032
15033
15034
15035
15036
15037
15038
15039
15040
15041
15042
15043
15044
15045
15046
15047
15048
15049
15050
15051
15052
15053
15054
15055
15056
15057
15058
15059
15060
15061
15062
15063
15064
15065
15066
15067
15068
15069
15070
15071
15072
15073
15074
15075
15076
15077
15078
15079
15080
15081
15082
15083
15084
15085
15086
15087
15088
15089
15090
15091
15092
15093
15094
15095
15096
15097
15098
15099
15100
15101
15102
15103
15104
15105
15106
15107
15108
15109
15110
15111
15112
15113
15114
15115
15116
15117
15118
15119
15120
15121
15122
15123
15124
15125
15126
15127
15128
15129
15130
15131
15132
15133
15134
15135
15136
15137
15138
15139
15140
15141
15142
15143
15144
15145
15146
15147
15148
15149
15150
15151
1515

this is the step of scraping our images from link

In [0]:
image_number=[]
for i in range(0,15634):
  image_number.append(str(i)+'.jpg')

  

here we make a coloumn of that we have number in our image id, in our image we have number from 1 to 15635, so we  make a loop from 1 to 13634 and convert in to dataframe and append in our dataset

In [0]:
image_number=pd.DataFrame(image_number)

In [0]:
dataset2=pd.concat([dataset2,image_number], axis=1)

In [0]:
dataset2.tail()


_unit_id   category  category:confidence  \
15629  851567928      plain               0.9232   
15630  851568211     animal               0.7667   
15631  854497512    squares               0.9230   
15632  854497595     tribal               0.5199   
15633  856388654  polka dot               0.9618   

                                               image_url          0  
15629  http://s3-eu-west-1.amazonaws.com/we-attribute...  15629.jpg  
15630  http://s3-eu-west-1.amazonaws.com/we-attribute...  15630.jpg  
15631  http://s3-eu-west-1.amazonaws.com/we-attribute...  15631.jpg  
15632  http://s3-eu-west-1.amazonaws.com/we-attribute...  15632.jpg  
15633  http://s3-eu-west-1.amazonaws.com/we-attribute...  15633.jpg

In [0]:
dataset2=dataset2.rename(columns={0:'image_number'})

here we do renaming the coloumn

In [0]:
dataset2.head()

_unit_id   category  category:confidence  \
0  851505458       ikat               0.3487   
1  851505459      plain               1.0000   
2  851505460  polka dot               0.6709   
3  851505461      plain               1.0000   
4  851505462   geometry               0.7035   

                                           image_url image_number  
0  http://s3-eu-west-1.amazonaws.com/we-attribute...        0.jpg  
1  http://s3-eu-west-1.amazonaws.com/we-attribute...        1.jpg  
2  http://s3-eu-west-1.amazonaws.com/we-attribute...        2.jpg  
3  http://s3-eu-west-1.amazonaws.com/we-attribute...        3.jpg  
4  http://s3-eu-west-1.amazonaws.com/we-attribute...        4.jpg

In [0]:
category=dataset2.iloc[:,1].values


In [0]:
dummy_categories=pd.get_dummies(category)

Here we create dummy variable of our category coloumn and append these dummy variable in our dataset

In [0]:
dummy_categories.shape

(15634, 17)

In [0]:
Newdataset=pd.concat([dataset2,dummy_categories],axis=1)

In [0]:
Newdataset.tail()

_unit_id   category  category:confidence  \
15629  851567928      plain               0.9232   
15630  851568211     animal               0.7667   
15631  854497512    squares               0.9230   
15632  854497595     tribal               0.5199   
15633  856388654  polka dot               0.9618   

                                               image_url image_number  OTHER  \
15629  http://s3-eu-west-1.amazonaws.com/we-attribute...    15629.jpg      0   
15630  http://s3-eu-west-1.amazonaws.com/we-attribute...    15630.jpg      0   
15631  http://s3-eu-west-1.amazonaws.com/we-attribute...    15631.jpg      0   
15632  http://s3-eu-west-1.amazonaws.com/we-attribute...    15632.jpg      0   
15633  http://s3-eu-west-1.amazonaws.com/we-attribute...    15633.jpg      0   

       animal  cartoon  chevron  floral   ...    ikat  letter_numb  plain  \
15629       0        0        0       0   ...       0            0      1   
15630       1        0        0       0   ...       0            0      0   
15631       0        0        0       0   ...       0            0      0   
15632       0        0        0       0   ...       0            0      0   
15633       0        0        0       0   ...       0            0      0   

       polka dot  scales  skull  squares  stars  stripes  tribal  
15629          0       0      0        0      0        0       0  
15630          0       0      0        0      0        0       0  
15631          0       0      0        1      0        0       0  
15632          0       0      0        0      0        0       1  
15633          1       0      0        0      0        0       0  

[5 rows x 22 columns]

In [0]:
Newdataset=Newdataset.drop(['_unit_id','category:confidence','image_url'],axis=1)

drop unit_id , category confidence, image url cooumn from our dataset

In [0]:
Newdataset.isnull().sum()

category        0
image_number    0
OTHER           0
animal          0
cartoon         0
chevron         0
floral          0
geometry        0
houndstooth     0
ikat            0
letter_numb     0
plain           0
polka dot       0
scales          0
skull           0
squares         0
stars           0
stripes         0
tribal          0
dtype: int64

In [0]:
Newdataset.tail()

category image_number  OTHER  animal  cartoon  chevron  floral  \
15629      plain    15629.jpg      0       0        0        0       0   
15630     animal    15630.jpg      0       1        0        0       0   
15631    squares    15631.jpg      0       0        0        0       0   
15632     tribal    15632.jpg      0       0        0        0       0   
15633  polka dot    15633.jpg      0       0        0        0       0   

       geometry  houndstooth  ikat  letter_numb  plain  polka dot  scales  \
15629         0            0     0            0      1          0       0   
15630         0            0     0            0      0          0       0   
15631         0            0     0            0      0          0       0   
15632         0            0     0            0      0          0       0   
15633         0            0     0            0      0          1       0   

       skull  squares  stars  stripes  tribal  
15629      0        0      0        0       0  
15630      0        0      0        0       0  
15631      0        1      0        0       0  
15632      0        0      0        0       1  
15633      0        0      0        0       0

In [0]:
Newdataset.columns

Index(['category', 'image_number', 'OTHER', 'animal', 'cartoon', 'chevron',
       'floral', 'geometry', 'houndstooth', 'ikat', 'letter_numb', 'plain',
       'polka dot', 'scales', 'skull', 'squares', 'stars', 'stripes',
       'tribal'],
      dtype='object')

In [0]:
columns=['OTHER', 'animal', 'cartoon', 'chevron',
       'floral', 'geometry', 'houndstooth', 'ikat', 'letter_numb', 'plain',
       'polka dot', 'scales', 'skull', 'squares', 'stars', 'stripes',
       'tribal']

In [0]:
model = Sequential()
model.add(Conv2D(32, (3, 3), padding='same',
                 input_shape=(28,28,3)))
model.add(Activation('relu'))
model.add(Conv2D(32, (3, 3)))
model.add(Activation('relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))
model.add(Conv2D(64, (3, 3), padding='same'))
model.add(Activation('relu'))
model.add(Conv2D(64, (3, 3)))
model.add(Activation('relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(512))
model.add(Activation('relu'))
model.add(Dropout(0.5))
model.add(Dense(17, activation='softmax'))
model.compile(optimizers.Adam(lr=0.001, beta_1=0.9, beta_2=0.999, epsilon=None, decay=0.0, amsgrad=False),loss="categorical_crossentropy",metrics=["accuracy"])

Instructions for updating:
Colocations handled automatically by placer.
Instructions for updating:
Please use `rate` instead of `keep_prob`. Rate should be set to `rate = 1 - keep_prob`.


In [42]:
test_datagen=ImageDataGenerator(rescale=1./255.)

train_generator=train_datagen.flow_from_dataframe(
dataframe=Newdataset[:13984],
directory="colabData/greendeck",
x_col="image_number",
y_col=columns,
batch_size=32,
seed=42,
shuffle=True,
class_mode="other",
target_size=(28,28))

valid_generator=test_datagen.flow_from_dataframe(
dataframe=Newdataset[13984:],
directory="colabData/greendeck",
x_col="image_number",
y_col=columns,
batch_size=32,
seed=42,
shuffle=True,
class_mode="other",
target_size=(28,28))



Found 13984 images.
Found 1650 images.


here we can see that 13984 images in training data that available in our drive, and 1650 images in our test data 

In [0]:
STEP_SIZE_TRAIN=train_generator.n//train_generator.batch_size
STEP_SIZE_VALID=valid_generator.n//valid_generator.batch_size
STEP_SIZE_TEST=test_generator.n//test_generator.batch_size
model.fit_generator(generator=train_generator,
                    steps_per_epoch=STEP_SIZE_TRAIN,
                    validation_data=valid_generator,
                    validation_steps=STEP_SIZE_VALID,
                    epochs=25)

for fitting model in our image dataset.Here we take batch size of 32. its accuracy may be 70 percent on unknown dataset. it may be because it breaking on during training  on training dataset due to breaking of image dataset.it break on 2 epoch. but i analyse that in first epoch accuracy on training dataset accuracy is 54% but on test dataset accuracy is 56% , so in each epoch error rate deacreases and accuracy increases,here i use epoch 25 so as we know at iteration of epoch accuracy increase.